sporkfish/endgame_tablebase.py

```diff
@@ -130,17 +133,26 @@ def query(self, board: Board) -> Optional[chess.Move]:
                 # This is a small performace hit but is miniscule compared to searching
                 cboard = chess.Board()
                 cboard.set_fen(board.fen())
-                for move in cboard.legal_moves:
+                dtz_scores = []
+                moves = list(cboard.legal_moves)
+                for move in moves:
                     cboard.push(move)
-                    # WDL Score: returns 2 if the side to move is winning, 0 if the position is a draw and -2 if the side to move is losing.
-                    # Returns 1 in case of a cursed win and -1 in case of a blessed loss.
-                    # Mate can be forced but the position can be drawn due to the fifty-move rule.
-                    # We check WDL score < 0 here, since we pushed our legal move and so its the opponents turn.
+                    # Returns a positive value if the side to move is winning, 0 if the position is a draw,
+                    # and a negative value if the side to move is losing.
+                    # DTZ = Number of moves to mate.
+                    # We check DTZ score < 0 here, since we pushed our legal move and so its the opponents turn.
                     # If they are losing, we are winning.
-                    if self._db.probe_wdl(cboard) < 0:
-                        return move
+                    dtz_score = self._db.probe_dtz(cboard)
+                    dtz_scores.append(dtz_score if dtz_score else float("inf"))
```

Comment:
@michelle-lai-swl on second thought I think we probably don't need random chance. By definition, DTZ is the number of moves to guaranteed mate (for simplicity). If on first probe we have a move such that DTZ = n , that means on the second probe we are guaranteed a move such that DTZ <= n -1. 

Suppose otherwise, on first probe we have a move such that DTZ = n, and on second probe all moves are of DTZ >= n. But then the first probe must have DTZ >= n + 1, since we already made a move. This contradicts our original assumption...

If you agree, I'll change to just use incremental updates (taking the least DTZ each time).

Reply:
Actually python-chess agrees with me https://python-chess.readthedocs.io/en/latest/syzygy.html:

I'll make the change.

> Minmaxing the DTZ50’’ values guarantees winning a won position (and drawing a drawn position), because it makes progress keeping the win in hand.

Reply:
Discussed offline, according to the documentation of DTZ, there can be some rounding issue. So DTZ can repeat itself after even an optimal move.

But I think the strategy of keep minimising opponents (-ve) DTZ should work, I.e. guarantee winning a winning board, or drawing a draw board.

But best to test...

Reply:
As @ccjeremylo found this *should* be okay because the databases we found are such that

> The tables up to 6 pieces on this site have been generated with [a patch to avoid rounding for 100 ≥ n ≥ 1](https://github.com/niklasf/tb/commit/9f1056429ad379eff52e748dc2783ea750671730).

Reply:
Actually it says nothing about -100 <= n <= -1.

Reply:
What is the conclusion here then?

Reply:
No - if you have any practical suggestions on how to prove the above that then open. Proof by counterexample is going to be a long search...

Reply:
Agree with you (leaving this for future....)